README.markdown
# pyJasper

pyJasper is a set of python based tools to handle JasperReports.
Since jasper reports is a Java application you need Java installed.

Jython (bundled with pyJasper) is used to access the JasperReports
library via HTTP. Your pure Python clients can transparently generate
JasperReport Documents.

See [StackOverflow](http://stackoverflow.com/questions/458340/is-there-a-better-layout-language-than-html-for-printing/459352#459352) and [this Blogposting](http://blogs.23.nu/c0re/2008/07/antville-18473/) to understand what problem pyJasper is trying to solve.

## Usage

### Servlet Interface

The servlet keeps no state at all. You have to supply it with an XML datasource, an XPath expression for that datasource and the JRXML report design. You get back the generated PDF or an plain text error message. The respective data has to be submitted via the form variables 'xpath', 'design' and 'xmldata'.

To try it out you can use curl. E.g. do to pyjasper/backend and start the jetty servlet container (`sh pyJasper-httpd.sh`). Then use  `curl` in another window to request rendering of a PDF.

    curl -X POST --form xpath=//lieferscheine/lieferschein 
                 --form design=@reports/Lieferschein.jrxml 
                 --form xmldata=@sample-xml/Lieferschein.xml 
                 --form sign_keyname={{ alias of the certificate in the KeyStore }} #optional
                 --form sign_reason={{ reason to be send w/ certificate }} #optional
                 --form callback={{ callback_URL }} #optional
                 http://localhost:8080/pyJasper/jasper.py > test.pdf

test.pdf should now contain a rendered PDF document.
If you provided sign_keyname and sign_reason AND your backend installation knows this keyname, you get a signed pdf.

With parameter callback you tell the server to generate the pdf and send it to the given URL instead of immediate return. 

To stop "the jetty servlet container" you can use the `sh pyJasper-httpd_stop.sh`

### Python interface

You are expected to subclass `pyjasper.JasperGenerator` and call it's `generate_pdf()` function. Usually you only have to overwrite the `__init__()` and `generate_xml(self, ...)` functions and use the the Python [ElementTree](http://docs.python.org/lib/module-xml.etree.ElementTree.html) API to generate an xml-tree. E.g.

    class MyPDFGenerator(JasperGenerator):
        """Jasper-Generator for Greetingcards"""
        def __init__(self):
            super(MovementGenerator, self).__init__()
            self.reportname = 'reports/Greeting.jrxml'
            self.xpath = '/greetings/greeting'
            self.root = ET.Element('gretings') 
    
        def generate_xml(self, tobegreeted):
            """Generates the XML File used by Jasperreports"""
            
            ET.SubElement(self.root, 'generator').text = __revision__
            for name in tobegreeted:
                xml_greeting  =  ET.SubElement(self.root, 'greeting')
                ET.SubElement(xml_greeting, "greeting_to").text = unicode(name)
                ET.SubElement(xml_greeting, "greeting_from").text = u"Max"
            return xmlroot

Now you can use `MyPDFGenerator` like this:

    generator = MyPDFGenerator()
    pdf = generator.generate(['nik', 'tobias', 'chris', 'daniel'])
    open('/tmp/greetingcard.pdf', 'w').write(pdf)

The Python client finds the URL of the Jasper Servlet by checking the @PYJASPER_SERVLET_URL@ environment variable. It this Variable is not set, a default value of `http://localhost:8080/pyJasper/jasper.py` is used.

### Installation

Get it at the [Python Cheeseshop](http://pypi.python.org/pypi/pyJasper) or at [GitHub](http://github.com/hudora/pyJasper)
To install the Python client interface  just execute `python setup.py install` as administrator. This should install the requred dependency [httplib2](http://code.google.com/p/httplib2/) automatically. For the Server part there exist no automatic setup script. Just copy `pyjasper/backend/`  to a suitable location and start `pyJasper-httpd.sh` I use Dan Bernsteins [supervise](http://cr.yp.to/daemontools/supervise.html) tool for running the Jetty server.

#### Use more Docker !!!

If you want, we are on docker hub: https://hub.docker.com/r/mportela/pyjasper/

so you can just pull the image:
`docker pull mportela/pyjasper`

Also, we have a self explanatory 'dockerfile', just build the container with `docker build` and later you can RUN it... this will start the servlet container in port "5555" so it's better to run your docker container with the `-p 5555:5555` to expose the servelt port

Example running the servlet with one line and without installing any deppendencies (just docker)

`docker run --rm -p 5555:5555 mportela/pyjasper`

#### exampledeployment on Amazon EC2

##### Create an new instance via AWS Management console

* Launch instance
  * Choose an AMI
    * Community AMI
    * e.g. ubuntu-images-eu/ubuntu-lucid-10.04-i386-server-20100923.manifest.xml
* Instance details
* Create Keypair
  * Create a new keypair or choose one of your existing
* Firewall
  * Custom: TCP Port 5555 - 5555
  * This custom option is not choosable in this setup-wizard but you can change the firewall settings later

##### Install and configure pyJasper

    ssh -A ubuntu@dns-name-of-instance.com
    sudo aptitude update
    sudo aptitude safe-upgrade
    sudo aptitude install openjdk-6-jre daemontools daemontools-run git-core
    cd /usr/local
    sudo git clone https://github.com/hudora/pyJasper.git
    cd /usr/local/pyJasper

Now you have to add a command to `/etc/crontab` to delete old files from `/tmp/pyJasper/` to keep the temporary files small. For example you can use this line to delete all files older than 7 days in `/tmp/pyJasper/`.

    0  1    * * *   root    /usr/bin/find /tmp/pyJasper/ -mtime +7 -type f -exec /bin/rm {} \;

Optional: Now you have to upload a keystore file for signing your documents and set it in `/usr/local/pyJasper/pyjasper/backend/pyJasper-httpd.sh`

    export PYJASPER_KEYSTORE_PASSWORD="thesecretpassphrase"
    export PYJASPER_KEYSTORE_FILE="/usr/local/pyJasper/signature-keystore-file.ks"

Finally configure daemontools to run the following command as a service:

    /usr/local/pyJasper/pyjasper/backend/pyJasper-httpd.sh -Xms128m -Xmx1G


## History

* **6.0.0*** bump version to reflect JasperReport version compatibility
* **0.2.1*** public release (Summer 2008)
* **0.2*** release based on Jetty/Servlets (late 2007)
* **0.1.1*** public release on hosted-projects.com
* **0.1*** release based on long running Java Process (late 2006)
* **.01** one Java process per document (2006)

## Files

* client.py                      - contains high-level Python functions for report generation. Should theoretically be able to work with report generators other than JasperReports. So far builds on JasperClient.py.
* backend/                       - contains tools to drive JasperReports and actually build reports. The backend was once based on XmlJasperInterface by Jonas Schwertfeger but hase benn moving away from it for some time.
* backend/lib                    - JasperReports and Jython Java Libraries
* reports                        - report source files (*.jrxml)
* sample-xml                     - sample XML files to work with reports
* sample-pdf                     - sample PDFs generated by using reports with sample XML

## Links

* [JasperReports](http://en.wikipedia.org/wiki/JasperReports)
* [How to Integrate JasperReports with Ruby on Rails](http://wiki.rubyonrails.org/rails/pages/howtointegratejasperreports)
* [jasapp](http://www.vmware.com/appliances/directory/311) a VMWare Virtuall Appliance providing a JasperReports Interface
* [Geraldo](http://geraldo.sourceforge.net/)


CHANGES
6.0.0 upgrade to JasperReports 6.0.0, added dockerfile (thanks https://github.com/mix86/pyJasper)
0.42 JasperGenerator.get_report gives an opportunity to override the loading of templates
0.41b comes with debugging in JasperGenerator
0.41 added get_reportname function
0.4 update to JasperReports 3.7 (thanks http://github.com/tomasd#)


pyjasper/client.py
#!/usr/bin/env python
# encoding: utf-8

"""
pyJasper client.py - Way to access JasperReports from Python using http.
This should be able to run under CPython, IronPython, PyPy, Jython or whatever.
See https://cybernetics.hudora.biz/projects/wiki/pyJasper for further enligthenment.

Created by Maximillian Dornseif on 2007-10-12.
Consider it BSD licensed.
"""

import os
import os.path
import uuid
import urllib
import xml.etree.ElementTree as ET
from cStringIO import StringIO
from httplib2 import Http


class JasperException(RuntimeError):
    """This exception indicates Jasper Server problem."""
    pass


def encode_multipart_formdata(fields):
    """
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be uploaded as files
    Return (content_type, body) ready for httplib.HTTP instance
    """
    # Based on From http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/146306
    boundary = '----------ThIs_Is_tHe_bouNdaRY_$%s' % (uuid.uuid4())
    out = []
    # if it's dict like then use the items method to get the fields
    if hasattr(fields, "items"):
        fields = fields.items()
    for (key, value) in fields:
        out.append('--' + boundary)
        out.append('Content-Disposition: form-data; name="%s"' % key)
        out.append('')
        out.append(value)
    out.append('--' + boundary + '--')
    out.append('')
    body = '\r\n'.join([str(x) for x in out])
    content_type = 'multipart/form-data; boundary=%s' % boundary
    return content_type, body


def get_reportname(base, *args):
    """
    Construct path for report file relative to base

    In most cases, this will be JasperGenerator.__file__
    """
    path = os.path.join(os.path.dirname(base), 'reports', *args)
    return os.path.abspath(path)


class JasperClient(object):
    """Generation of JasperReport documents using the Jetty/Jython Servelet."""

    def find_server_url(self):
        """Return the URL of the page where the server lives.

        Checks the PYJASPER_SERVLET_URL default variable."""

        return os.getenv('PYJASPER_SERVLET_URL', default='http://localhost:8080/pyJasper/jasper.py')

    def generate_server(self, design, xpath, xmldata, metadata, multi=False, output_type='pdf'):
        """Generate report via pyJasperServer."""

        url = self.find_server_url()
        
        fields = dict(design=design, xpath=xpath, xmldata=xmldata, output_type=output_type)
        if metadata:
            fields['metadata'] = urllib.urlencode(metadata)

        if multi:
            content_type, content = encode_multipart_formdata(fields=fields)
        else:
            content_type, content = encode_multipart_formdata(fields=fields)

        resp, content = Http().request(url, 'POST', body=content, headers={"Content-Type": content_type})
        if not resp.get('status') == '200':
            raise JasperException("%s -- %r" % (content, resp))
        return content

    def generate_report(self, design, xpath, xmldata, metadata, multi=False, output_type='pdf'):
        """Generate report via pyJasperServer."""
        return self.generate_server(design, xpath, xmldata, metadata, multi, output_type=output_type)


class JasperGenerator(object):
    """Abstract class for generating Documents out with Jasperreports.

    You have to overwrite generate_xml to make meaningfull use of this class. Then call
    YourClass.generate(yourdata). Yourdata is passed to generate_xml() and hopfully you will get
    the generated report back.
    """

    def __init__(self, debug=False):
        super(JasperGenerator, self).__init__()
        self.reportname = None
        self.xpath = None
        self.debug = debug
        self.metadata = None

    def generate_xml(self, data=None):
        """To be overwritten by subclasses.

        E.g.
        def generate_xml(self, movement):
            ET.SubElement(self.root, 'generator').text = __revision__
            ET.SubElement(self.root, 'generated_at').text = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
            xml_movement  =  ET.SubElement(xmlroot, 'movement')
            ET.SubElement(xml_movement, "location_from").text = unicode(movement.location_from)
            return xmlroot
        """
        raise NotImplementedError

    def get_xml(self, data=None):
        """Serializes the XML in the ElementTree to be send to JasperReports."""
        root = self.generate_xml(data)
        tree = ET.ElementTree(root)
        buf = StringIO()
        tree.write(buf, encoding="utf-8")
        ret = buf.getvalue()
        buf.close()
        return ret

    def get_report(self):
        """Get JasperReport template"""
        report = open(self.reportname)
        res = report.read()
        report.close()
        return res


    def generate_report(self, data=None, output_type='pdf'):
        """Generates a document by using Jasper-Reports."""
        server = JasperClient()
        design = self.get_report()
        xmldata = self.get_xml(data)
        if self.debug:
            open('/tmp/pyjasper-%s-debug.xml' % os.path.split(self.reportname)[-1], 'w').write(xmldata)

        return server.generate_report(design, self.xpath, xmldata, self.metadata, output_type=output_type)


    def generate(self, data=None, output_type='pdf'):
        """Generates a report, returns the PDF or XLS."""

        if output_type not in ('pdf', 'xml', 'rtf', 'xls', 'csv', 'text', 'html'):
            raise JasperException, 'Unsupported data_type %s. Only pdf and xls supported.'

        return self.generate_report(data, output_type=output_type)



pyjasper/client_subreport.py
#!/usr/bin/env python
# encoding: utf-8
"""
pyJasper clientSubreport.py - Way to access JasperReports from Python using http when using subReports.
This should be able to run under CPython, IronPython, PyPy, Jython or whatever when (simple)json is available.
See https://cybernetics.hudora.biz/projects/wiki/pyJasper for further enligthenment.

Created by Johan Otten on 2009-09-24.
Consider it BSD licensed.
"""

import xml.etree.ElementTree as ET
from cStringIO import StringIO
import os.path
import simplejson as json


from client import JasperClient


class JasperGeneratorWithSubreport(object):
    """Abstract class for generating Documents with subreports out with Jasperreports with.

    You have to overwrite generate_xml to make meaningfull use of this class. Then call
    YourClass.generate(yourdata). Yourdata is passed to generate_xml() and hopfully you will get
    the generated report back.
    """

    def __init__(self):
        super(JasperGeneratorWithSubreport, self).__init__()

        self.metadata = None
        self.mainreport = os.path.abspath(os.path.join(self.reportrootdir, self.reportbase + '.jrxml'))
        #self.mainreport = None
        self.subreportlist = []

        for filename in os.listdir(os.path.abspath(self.reportrootdir)):
            if filename.startswith(self.reportbase + '-subreport'):
                self.subreportlist.append(os.path.join(self.reportrootdir, filename))

        self.xpath = None

    def generate_xml(self, data=None):
        """To be overwritten by subclasses.

        E.g.
        def generate_xml(self, movement):
            ET.SubElement(self.root, 'generator').text = __revision__
            ET.SubElement(self.root, 'generated_at').text = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
            xml_movement  =  ET.SubElement(xmlroot, 'movement')
            ET.SubElement(xml_movement, "location_from").text = unicode(movement.location_from)
            return xmlroot
        """
        raise NotImplementedError

    def get_xml(self, data=None):
        """Serializes the XML in the ElementTree to be send to JasperReports."""
        root = self.generate_xml(data)
        tree = ET.ElementTree(root)
        buf = StringIO()
        tree.write(buf, encoding="utf-8")
        ret = buf.getvalue()
        buf.close()
        return ret

    def generate_pdf(self, data=None):
        """Generates a PDF document by using Jasper-Reports."""
        server = JasperClient()
        designs = {'main': open(self.mainreport).read()}
        multi = False
        for report in self.subreportlist:
            multi = True
            name = str(os.path.splitext(os.path.basename(report))[0])
            designs[name] = open(report).read()

        reportjson = json.dumps(designs)

        xmldata = self.get_xml(data)
        return server.generate_pdf(reportjson, self.xpath, xmldata, self.metadata, multi=multi)

    def generate(self, data=None):
        """Generates a report, returns the PDF."""
        return self.generate_pdf(data)


pyjasper/backend/pyJasper-httpd.sh
#!/bin/sh

# This starts the jetty based http server for pyJasper. This server can be
# used to generate Reports with JasperReport. The server keeps no state at
# all. You have to supply it with an XML datasource, an XPath expression
# for that datasource and the JRXML report design. You get back the
# generated PDF or an plain text error message.

# To try it out you can use curl. E.g.
# curl -X POST --form xpath=//lieferscheine/lieferschein \
#              --form design=@reports/Lieferschein.jrxml \
#                \
#              http://localhost:8080/pyJasper/jasper.py  > test.pdf

# go to script dir
cd `dirname $0`

# This is needed to work properly with unicode on Solaris 10
LC_CTYPE=en_US.UTF-8
export LC_CTYPE

# generate classpath of all all JARs in lib and the fonts dir
MYCLASSPATH=`echo lib/*.jar|sed 's/ /:/g'`:fonts:.
PATH=$PATH:/usr/local/bin
mkdir -pv logs

# keytool -genkey -keyalg RSA \
#         -alias hudora-rechnungen \
#         -keypass pw \
#         -keystore keystore.ks \
#         -dname "cn=HUDORA Rechnungen"
export PYJASPER_KEYSTORE_PASSWORD="topSecrect"
export PYJASPER_KEYSTORE_FILE="/usr/local/pyJasper/signature-keystore.ks"

java $@ -cp $MYCLASSPATH -Dfile.encoding=utf-8 \
  -Dnet.sf.jasperreports.awt.ignore.missing.font=true \
  -Dpython.home=./lib \
  -Djetty.port=5555 \
  -DSTOP.KEY=blaat -DSTOP.PORT=8079 \
  -Djava.awt.headless=true \
  -jar start.jar


pyjasper/backend/webapps/pyJasper/jasper.py
#!/usr/bin/env python
# encoding: utf-8
"""
jasper.py - pyJasper Servlet

Created by Maximillian Dornseif on 2007-09-09.
Copyright (c) 2007 HUDORA GmbH. All rights reserved.
"""

import sys
sys.path.extend(['.'])

import java.lang
from javax.servlet.http import HttpServlet
from org.apache.commons.fileupload.servlet import ServletFileUpload
from org.apache.commons.fileupload.disk import DiskFileItemFactory
import XmlJasperInterface

import cgi
import simplejson
import threading
import urllib
import urllib2


CONTENT_TYPES = {
    'pdf': "application/pdf",
    'xml': "application/xml",
    'rtf': "application/rtf",
    'xls': "application/vnd.ms-excel",
    'csv': "text/plain",
    'text': "text/plain",
    'html': "text/html",
}


def respond(data):
    """Write response to a given callback"""

    headers = {"Content-type": CONTENT_TYPES[data['output_type']]}
    stream = generate(data)

    try:
        request = urllib2.Request(data['callback'], stream, headers)
        urllib2.urlopen(request)
    except urllib2.URLError:
        pass


def generate(data):
    """Get rendered report from JasperReports"""

    jaspergenerator = XmlJasperInterface.JasperInterface(data['designs'], data['xpath'])
    return jaspergenerator.generate(data['xmldata'], output_type=data['output_type'] or 'pdf',
                                    sign_keyname=data['sign_keyname'], sign_reason=data['sign_reason'],
                                    metadata=data['metadata'])


class jasper(HttpServlet):
    """Servlet for rendering JasperReport reports.

    The servlet keeps no state at all. You have to supply it with an
    XML datasource, an XPath expression for that datasource and
    the JRXML report design. You get back the generated PDF or an
    plain text error message.

    The respective data has to be submitted via the form variables
    'xpath', 'design' and 'xmldata'.

    To try it out you can use curl. E.g.
    curl -X POST --form xpath=//lieferscheine/lieferschein
                 --form design=@reports/Lieferschein.jrxml
                 --form xmldata=@sample-xml/Lieferschein.xml
                 http://localhost:8080/pyJasper/jasper.py > test.pdf
    """

    def doGet(self, request, response):
        """Handle GET requests by just redirecting to doPost()."""
        # we use the same handler for GET and POST. I wonder:
        # actually GET is correct since there is no state change on the server
        # but POST can handle more data
        self.doPost(request, response)

    def doPost(self, request, response):
        """Generates a PDF with JasperReports. To be called with:

           xmldata: XML data Source for JasperReports
           xpath:   XPATH Expression for selecting rows from the datasource
           design:  JasperReports JRXML Report Design
           designs: JasperReports JRXML Report Design when using subreports.
           sign_keyname: Keyname for signing PDF files - this parameter is optional.
           sign_reason: Reason to be send w/ signature
           metadata: PDF metadata (see http://www.pdfa.org/doku.php?id=artikel:en:pdfa_metadata)
        """

        data = {'xpath': request.getParameter('xpath'),
                'design': request.getParameter('design'),
                'designs': request.getParameter('designs'),
                'xmldata': request.getParameter('xmldata'),
                'sign_keyname': request.getParameter('sign_keyname'),
                'sign_reason': request.getParameter('sign_reason'),
                'callback': request.getParameter('callback'),
                'metadata': request.getParameter('metadata'),
                'output_type': request.getParameter('output_type'),
               }

        if ServletFileUpload.isMultipartContent(request):
            servlet_file_upload = ServletFileUpload(DiskFileItemFactory())
            files = servlet_file_upload.parseRequest(request)
            fiterator = files.iterator()
            while fiterator.hasNext():
                file_item = fiterator.next()
                data[file_item.getFieldName()] = str(java.lang.String(file_item.get()))

        if data['designs']:
            data['designs'] = simplejson.loads(data['designs'])

        if not data['designs']:
            data['designs'] = {'main': data['design']}

        # Decode metadata
        if data['metadata']:
            metadata = {}
            for key, value in cgi.parse_qs(urllib.unquote(data['metadata'])).items():
                metadata[key] = value[0]
            data['metadata'] = metadata

        out = response.getWriter()
        if not data['xpath']:
            out.println('No valid xpath: %r\nDocumentation:' % data['xpath'])
            out.println(self.__doc__)
        elif not data['designs']:
            out.println('No valid design: %r\nDocumentation:' % data['designs'])
            out.println(self.__doc__)
        elif not data['xmldata']:
            out.println('No valid xmldata: %r\nDocumentation:' % data['xmldata'])
            out.println(self.__doc__)
        else:
            if data.get('callback'):
                writer = self.callback_response
            else:
                writer = self.immediate_response
            writer(request, response, data)

        out.close()

    def callback_response(self, request, response, data):
        """Return data by calling a webhook"""

        response.setContentType("text/plain")
        out = response.getWriter()
        out.println('Data will be send to %s' % data['callback'])

        thread = threading.Thread(target=respond, args=(data, ))
        thread.start()

    def immediate_response(self, request, response, data):
        """Return data immediatly"""
        output_type = data.get('output_type') or 'pdf'
        print output_type
        
        response.setContentType(CONTENT_TYPES[output_type])
        out = response.getWriter()
        data = generate(data)
        out.write(data)
